fix(fhir): SAV-847: Filter out unresolved resources from FHIR search API (hotfix 2.16.12)

rohan-bes opened the pull request:
### Changes

This fix is being implemented to avoid a bug that occurs when a client fetches a resource from our FHIR API before it has been fully resolved. In this scenario, they end up stuck with a resource where the IDs in its references can't be used to fetch the referenced resource.

Some core changes here to the way we resolve references in FHIR:
- Added a `resolved` flag to every resource
- When a resource gets resolved by the resolver, we update it to have resolved=true
- When a resource rematerialises, we check if its references exist and if so we keep them rather than setting to the original unresolved value (this is to avoid resources frequently 'disappearing' from the API if they kept getting marked as unresolved)
- When a resource is resolved, we bump its last_updated so that it can be refetched

**UPDATE**
So after running into issues with bumping the `last_updated` via the database, and feeling a bit uncomfortable about managing the resolution logic in multiple places, I decided to go with changing the logic to only use the javascript resolution logic and to drop the database functions. This ultimately seems like it'd be simpler in the long run, but it is a core change. I've left in a bunch of the migrations for now but will drop them before merging (just to keep a record and for discussion during code review). 

### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

Commit:
fix(fhir): SAV-847: Added 'resolved' column to fhir resource tables

Commit:
fix(fhir): SAV-847: Set resolved=true on fhir resources once resolver has run

Commit:
fix(fhir): SAV-847: Consolidate resolved/unresolved reference construction

Commit:
fix(fhir): SAV-847: Reworked building FhirReferences to avoid resolving references each time we rematerialise

Commit:
fix(fhir): SAV-847: Updated resolved when rematerialising a resource

Commit:
fix(fhir): SAV-847: Filter out unresolved resources from the fhir search API
- For now leaving them as GETable

Commit:
fix(fhir): SAV-847: Added bumping of last_updated to fhir resolver procedures

Commit:
fix(fhir): SAV-847: Fix missing display on practitioner references

Commit:
fix(fhir): SAV-847: Fix sorting in tests

Commit:
fix(fhir): SAV-847: Switched Resource.resolveUpstreams() to rely on rematerialisation rather than the fhir.resolve_upstreams db function

Commit:
fix(fhir): SAV-847: Add migration to remove fhir.resolve_upstreams database functions

Commit:
fix(fhir): SAV-847: Added unit tests for fhir resolver

Commit:
fix(fhir): SAV-847: Updated FhirSpecimen test to ensure all references were resolved

Commit:
fix(fhir): SAV-847: Fixed bug in specimen formatting on ServiceRequests

chris-bes reviewed (changes requested):
Functionality wise they all make sense to me and I don't see any implications! I just have a few minor suggestions 🙏

Commit:
fix(fhir): SAV-847: PR Fixups
- Consolidated unresolved reference type logic
- Fixed possible NPE
- Reordered imports

chris-bes reviewed (approved):
LGTM!

Commit:
fix(fhir): SAV-847: Require resources's references to be resolved before marking as resolved

Commit:
fix(fhir): SAV-847: Throw an error when fetching a single resource if not resolved within a timeout

Commit:
fix up sleepAsync import

Commit:
fix(fhir): SAV-847: Reworked patient merge tests to ensure all patients are first materialised

Commit:
fix(fhir): SAV-847: Added tests for await resolution logic

Commit:
fix(fhir): SAV-847: Added delay in resolver job and wrapped in transaction

Commit:
fix(fhir): SAV-847: Removed migrations that changed the resolver functions as we're just removing them anyway

Commit:
fix(fhir): SAV-847: Moved sleep outside of transaction

Commit:
Bumped wait to see if it fixes intermittent test failure

rohan-bes committed:
Merge branch 'release/2.16' into fix/sav-847-fhir-fix-unresolved-ids
